### PR TITLE
Add support for software text editing controls

### PR DESCRIPTION
**Pull request by edman**
Includes selection, copy, cut, paste, as well as partial support for up
and down movement.

Text editing controls can be accessed in GBoard by:
  top-left arrow > three dots menu > text editing

Partial fix for flutter/flutter#9419 and flutter/flutter#37371.

### CODE REVIEW

**Comment by edman** [comment]
Some comments about the limitations of this PR:

* Home/End are not supported. I'm not sure where events for those keys are sent in the engine.

* Up/Down work except when moving around long lines that wrap around the view (see the video above). Basically the InputConnectionAdaptor does not know the width of the text view and thinks text always fits.
https://github.com/flutter/engine/blob/72420a3ba5cfe69e3476df5e56e49dde56021f0f/shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java#L47-L49

File: shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java [hunk 1]
```diff
@@ -195,28 +198,128 @@ public boolean sendKeyEvent(KeyEvent event) {
...
             } else if (event.getKeyCode() == KeyEvent.KEYCODE_DPAD_LEFT) {
                 int selStart = Selection.getSelectionStart(mEditable);
-                int newSel = Math.max(selStart - 1, 0);
-                setSelection(newSel, newSel);
+                int selEnd = Selection.getSelectionEnd(mEditable);
+                if (selStart == selEnd && !event.isShiftPressed()) {
+                    int newSel = Math.max(selStart - 1, 0);
+                    setSelection(newSel, newSel);
+                }
+                else {
```
Comment (justinmc) [inline shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java]: Nit: I think the usual style in our Java files is to put this on the same line like `} else {`.  Also below.

File: shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java [hunk 2]
```diff
@@ -195,28 +198,128 @@ public boolean sendKeyEvent(KeyEvent event) {
...
+                int selStart = Math.max(0, Selection.getSelectionStart(mEditable));
+                int selEnd = Math.max(0, Selection.getSelectionEnd(mEditable));
+                int selMin = Math.min(selStart, selEnd);
+                int selMax = Math.max(selStart, selEnd);
+                int newSelStart = selMin + textToPaste.length();
+                if (selMin != selMax)
+                    mEditable.delete(selMin, selMax);
+                mEditable.insert(selMin, textToPaste);
+                setSelection(newSelStart, newSelStart);
+				    }
```
Comment (justinmc) [inline shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java]: Weird indentation here.

Comment (justinmc) [reply shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java]: Should there be a `return true;` for this case?

**Comment by GaryQian** [comment]
Obtaining the width at the embedder level is non-trivial. Unless it is urgently needed, I would hold off on developing a system to pass width down to there.

The sendKeyEvent method should be unit-testable?

**Comment by justinmc** [comment]
Alright let's leave the up/down behavior as-is for now and revisit if needed.

Unit testing sounds like a good idea.  @edman You'll have to add a new unit test file for this since one doesn't already exist for InputConnectionAdaptor unfortunately.  It should be somewhat similar to the unit test file added in this PR: https://github.com/flutter/engine/pull/12432/files#diff-af19d10c7a69b9cab3ee96c4f64a9a9cR1

File: shell/platform/android/test/io/flutter/plugin/editing/InputConnectionAdaptorTest.java [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+        KeyEvent upKeyDown = new KeyEvent(KeyEvent.ACTION_DOWN, KeyEvent.KEYCODE_DPAD_UP);
+        boolean didConsume = adaptor.sendKeyEvent(upKeyDown);
+
+        assertTrue(didConsume);
+        int expectedSelection = selStart - SAMPLE_TEXT.indexOf('\n');
+        // WIP
+        // The asserts below don't work. KEYCODE_DPAD_UP moves the caret to
+        // index 0, as if the text were in one single line.
+        // assertEquals(expectedSelection, Selection.getSelectionStart(editable));
+        // assertEquals(expectedSelection, Selection.getSelectionEnd(editable));
```
Comment (edman) [inline shell/platform/android/test/io/flutter/plugin/editing/InputConnectionAdaptorTest.java]: Couldn't get up/down to work as expected yet.

Up moves selection to beginning of the text and down moves to end. There's a line break in the text so it should move to the next/previous line instead.

Not sure why it behaves differently in tests.. Will look more tomorrow.

Comment (GaryQian) [reply shell/platform/android/test/io/flutter/plugin/editing/InputConnectionAdaptorTest.java]: So I have noticed that depending on the input method, some keystrokes/events do not pass through this method at all. I am not yet clear on where those events go but that behavior may have something to do with this.

**Comment by justinmc** [comment]
We should check whether this fixes https://github.com/flutter/flutter/issues/50875#issuecomment-587761919.

**Comment by gschier** [comment]
EditableText contains its own logic for arrow navigation of hardware keyboards, including up/down. Could do the same for software keyboard too

**Comment by edman** [comment]
@justinmc I changed the test logic and added a note.

**Comment by justinmc** [comment]
@gschier Thanks for pointing that out, I hadn't realized it was being handled in both places.  I talked to @gspencergoog and it sounds like it's due for a refactor to consolidate.  This should be fine for now.

**Comment by justinmc** [comment]
Hmm I reran that failure and it succeeded, but it's still red.  @edman Can you try merging with master and pushing?

**Comment by edman** [comment]
@GaryQian FYI we'll be landing this once the tree is green.

**Pull request merged**